### PR TITLE
Add scale in controller that periodically scales down

### PR DESCRIPTION
**Pull request by hasit**
### Description of change

Screenshot from 2023-05-11 10-39-20

Screenshot from 2023-05-11 10-39-29

##### Checklist

- [x] Tested in playground or other setup
- [x] Screenshot (Grafana) from playground added to PR for 15+ minute run
- [x] Documentation is changed or added
- [ ] Tests and/or benchmarks are included
- [ ] Breaking changes

### Summary by OpenAI

**New Feature:**
- Added `PeriodicDecrease` controller for auto-scaling based on a periodic timer
- Modified `ScaleInController` message to include the new `PeriodicDecrease` controller

> 🎉 A timer ticks, replicas shrink,
> With PeriodicDecrease, we sync.
> Scaling down, a rhythm found,
> In this PR, new features abound! 🚀

### CODE REVIEW

File: api/aperture/policy/language/v1/autoscale.proto [hunk 1]
```diff
@@ -134,6 +134,18 @@ message DecreasingGradient {
   Parameters parameters = 2;
 }
 
+// PeriodicDecrease defines a controller for scaling in based on a periodic timer.
+message PeriodicDecrease {
+  message Parameters {
+    // The period of the timer.
+    google.protobuf.Duration period = 1; // @gotags: default:"1m"
+    // The percentage of replicas to scale in.
```
Comment (tanveergill) [inline api/aperture/policy/language/v1/autoscale.proto:142]: This controller is independent of Kubernetes so use a generic language like "percentage of scale to reduce"

Comment (hasit) [reply api/aperture/policy/language/v1/autoscale.proto]: Addressed in [0d12ad9](https://github.com/fluxninja/aperture/pull/1923/commits/0d12ad9a06acab38385913cff7753406989a3211)

File: pkg/policies/controlplane/components/autoscale/auto-scaler.go [hunk 2]
```diff
@@ -732,20 +746,75 @@ func ParseAutoScaler(
...
+										},
+									},
+								},
+							},
+							OutPorts: &policylangv1.Max_Outs{
+								Output: &policylangv1.OutPort{
+									SignalName: scaleInReductionPreCeilAdjusted,
 								},
 							},
 						},
```
[the remark this answers is not in the document]
Comment (hasit) [reply pkg/policies/controlplane/components/autoscale/auto-scaler.go:820]: That is very interesting. We currently do not have a Clamp component. I will make it in the near future.

File: api/aperture/policy/language/v1/autoscale.proto [hunk 3]
```diff
@@ -134,6 +134,18 @@ message DecreasingGradient {
   Parameters parameters = 2;
 }
 
+// PeriodicDecrease defines a controller for scaling in based on a periodic timer.
+message PeriodicDecrease {
```
[the remark this answers is not in the document]
Comment (harjotgill) [reply api/aperture/policy/language/v1/autoscale.proto:138]: Maybe doesn't matter as we use increase/decrease in the context of gradient controllers.

File: pkg/policies/controlplane/components/autoscale/auto-scaler.go [hunk 4]
```diff
@@ -758,39 +815,50 @@ func ParseAutoScaler(
...
+							},
+						},
+					},
+				},
+				{
+					Component: &policylangv1.Component_PulseGenerator{
+						PulseGenerator: &policylangv1.PulseGenerator{
+							FalseFor: parameters.Period,
+							TrueFor:  durationpb.New(policyReadAPI.GetEvaluationInterval()),
+							OutPorts: &policylangv1.PulseGenerator_Outs{
```
[the remark this answers is not in the document]
Comment (hasit) [reply pkg/policies/controlplane/components/autoscale/auto-scaler.go:859]: `TrueFor` requires duration in `*durationpb.Duration`

[the remark this answers is not in the document]
Comment (hasit) [reply pkg/policies/controlplane/components/autoscale/auto-scaler.go:859]: TrueFor requires duration in *durationpb.Duration

File: api/aperture/policy/language/v1/autoscale.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (hasit) [reply api/aperture/policy/language/v1/autoscale.proto:138]: Thought was to stick to the same spirit of `DecreasingGradient` and not involve `ScaleIn` into the name.

[the remark this answers is not in the document]
Comment (hasit) [reply api/aperture/policy/language/v1/autoscale.proto:138]: @harjotgill Following is what the user sees:

```
      - auto_scale:
          auto_scaler:
            scale_in_controllers:
              - controller:
                  periodic:
                    parameters:
                      period: 60s
                      scale_in_percentage: 10
            scale_out_controllers:
              - controller:
                  gradient:
                    in_ports:
                      setpoint:
                        constant_signal:
                          value: 1
                      signal:
                        signal_name: DESIRED_LOAD_MULTIPLIER
                    parameters:
                      slope: -1
```

Notice the `periodic` block in `scale_in_controllers`

**Review by tanveergill — APPROVED** [review]

**Pull request merged**